### DISCUSSION TITLE
Ethernet functional workaround

### DISCUSSION THREAD

**Discussion by PavelKuzub** [HomeSpan/HomeSpan#738] [General]
Hello Gregg, thanks for this wonderful library.

In my project for a garage door opener, I use [Silicognition wESP32](https://wesp32.com/) board for arguably best-in-class PoE Ethernet (isolated power circuits, high wattage, but comes at a higher cost).
Everything in the project has been Ethernet-oriented for a stable connection.
HomeKit integration has been done via MQTT to Home Assistant and then to HomeKit.
Connecting to HomeKit directly was a last-minute nice-to-have feature to remove the HA hop and so I have discovered your library.

I have seen other discussions about Ethernet support and I know you initially had problems with mDNS and then lack of time/priority to look into adding full support.
From my experience working with the wESP32 board - the mDNS broadcasting is working fine over ETH under ESP32 Arduino 2.0.5.

Your library explicitly wants to start WiFi and would only start hapServer and broadcast mDNS if WiFi credentials are set and the WiFi status is WL_CONNECTED. No public methods in HS 1.8.0 to alter this behavior. Tough to work with.

So I have been looking to find a way to bootstrap HomeSpan to Ethernet project without modifying HomeSpan code and just importing it as a dependency. I have succeeded.

I have come up with a dirty hack for now, though would love to see HomeSpan library expose some setter method that would not require WiFi credentials and WiFi connected state to start hapServer and broadcast mDNS.

Providing a demo code to illustrate the functional Ethernet-based operation of HomeSpan.

Looking forward to a feature flag in future versions to stop using this hack. Thank you. 
A bonus feature request is to expose a method to re-broadcast HAP mDNS, so it can be added to the ARDUINO_EVENT_ETH_GOT_IP event.

```
/* 
  Demo to run HomeSpan using Ethernet instead of WiFi  
  Pavel Kuzub
  2024-01-08
*/

#include <Arduino.h>
#include <WiFi.h>
#include <ETH.h>
#include <ESPmDNS.h>
#include <HomeSpan.h>

void gen_random(char *s, int l) {
    for (int c; c=rand()%62, *s++ = (c+"07="[(c+16)/26])*(l-->0););
}

void setup(){
  // Serial port for debugging purposes
  Serial.begin(115200);
  while (!Serial && millis() < 5000);

  // Start the Ethernet, wESP32 revision 7+ boards uses RTL8201
  ETH.begin(0, -1, 16, 17, ETH_PHY_RTL8201); 

  // To prevent HomeSpan from adding its suffix to the hostname set in other parts of the project
  homeSpan.setHostNameSuffix("");

  // The Setup code
  homeSpan.setPairingCode("46637726");
  homeSpan.setQRID("HSPN");

  // Change the port in case your project is using port 80 for something else
  homeSpan.setPortNum(8080);

  // HomeSpan requires having non-empty WiFi credentials to start. 
  // We will make it not connecting to WiFi at all, but as a precaution, will
  // Generate a random string for fake SSID and password, in case HomeSpan ever actually attempts to connect to WiFi.
  // There is not going to be a WiFi network with such a random SSID and such a random password.
  //Given that it is random - it would be impossible to spoof by creating such an Access Point solely by looking into this code, even if it will attempt to connect to such SSID.
  char FakeSSID[17];
  char FakePASS[17];
  gen_random(FakeSSID, 16);
  gen_random(FakePASS, 16);

  // For Debug - verify that these fake SSID and Password are truly random on each boot of the device.
  Serial.print("FakeSSID: ");
  Serial.println(FakeSSID);
  Serial.print("FakePASS: ");
  Serial.println(FakePASS);

  // ETH Hack part 1 - Set Fake WiFi credentials
  homeSpan.setWifiCredentials(FakeSSID,FakePASS);

  // ETH Hack part 2 - Set Fake status of WiFi to Connected, so HomeSpan is satisfied and does not attempt to start WiFi at all
  WiFi._setStatus(WL_CONNECTED);

  // Start HomeSpan
  homeSpan.begin(Category::Lighting, "HomeSpan LightBulb", "customhostname", "HomeSpan-ESP32");

  // Demo accessory
  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify();

  new Service::LightBulb();
    new Characteristic::On();

}

void loop(){
  homeSpan.poll();
}
```

**Comment by HomeSpan** [comment]
Thanks for this.  I actually have a branch of HomeSpan that works with Ethernet (as you noted, only a few small modifications needed), though it's quite outdated since I have not worked on it in a few years.  The problem I had was working with a Wiznet W5500 Featherboard from Adafruit (sitting on top an ESP32).  Worked perfectly except for MDNS.  Research on the web revealed that the problem was deep in the UDP driver for that chip.  However, drivers for other chips may not have that issue, and it looks like the w32 uses an RTL8201, which you're finding works well with MDNS.

This gives me a good reason to revisit my Ethernet version and refresh.  Do you know if the w32 is available for sale anywhere (it does not seem to be on Amazon)?

**Reply by PavelKuzub** [reply]
I have bought several [wESP32 boards from Amazon](https://www.amazon.com/dp/B084RV91NQ/), and some [wESP32 from Tindie](https://www.tindie.com/products/silicognition/wesp32/) where Silicognition LLC has larger variety of items and has stock. This solution comes at a premium as it also requires a separate prog board. Micro USB variant [wESP32-Prog](https://www.tindie.com/products/silicognition/wesp32-prog/) ($8 when ordered on wESP32 order form) and recent addition USB-C variant [wESP32-Prog-C](https://www.tindie.com/products/silicognition/wesp32-prog-c/)

Hope this board will recover your faith in Ethernet on ESP32. Thank you.

**Reply by HomeSpan** [reply]
Thanks - I've been digging more into Arduino-ESP32 code for how they expect ETH to be used, and must to my surprise, I did not realize that the ETH implementation is actually an overlay of the WiFi library.  This is why your method works even though the HomeSpan code uses WiFiClient and WiFiServer everywhere.  Even without a working ETH board I should be able to make some targeted modifications to the library to allow the user to drop in their own "initial network connectivity."  By coincidence another user just had a similar request - they were using WPA2-Enterprise to connect to their school's network and performed a similar "hack" to get HomeSpan to think it was connected.  Should be straightforward to add in a user-defined mechanism.

**Reply by Stuckya** [reply]
Reviving this thread, would also love for this to be natively supported. Any updates @HomeSpan?

**Reply by HomeSpan** [reply]
Yes, will be the main focus of the next release.

The problem I was having with MDNS seems to be related to the drivers Arduino uses for the Wiznet 5100.  If you use the Espressif ETH.h library, MDNS seems to work fine.

**Reply by 011V32** [reply]
Hi.
Any news on that?
In the new 2.0.0-rc.1 release there is only the "prepare for future integration of Ethernet support" information.
Would be great if this could be integrated.
Thx.

**Reply by HomeSpan** [reply]
Will be in release right after this one.

**Reply by chrmuh** [reply]
Hey,
can you tell when this next release will approximately be.
If you need people to test it before you release the library I would look forward to test it for you. I have a olimex-POE-ESP32 board.
Best wishes

**Comment by HomeSpan** [comment]
likely not for a few months.  Will let you know when ready for initial testing (much appreciated).  Espressif dramatically re-worked the networking code in Arduino-ESP32 3.X, so a good deal of refactoring will be done in HoneSpan to make best use of the new network features.

**Reply by chrmuh** [reply]
Ok, just let us now for testing. Is it possible to start projects wie WIFI and than just update the library, when it is ready?

**Comment by chrmuh** [comment]
Hey,

I try to redo this project. But the following line is not working. Do I need to add this function to the WiFi.h libary. In mine this privat function does not exist.

>   // ETH Hack part 2 - Set Fake status of WiFi to Connected, so HomeSpan is satisfied and does not attempt to start WiFi at all
>   WiFi._setStatus(WL_CONNECTED);

Which libary have you used?


Best wishes

**Comment by HomeSpan** [comment]
At long last, HomeSpan now has integrated support for Ethernet connectivity (I was waiting for Espressif to release their new Arduino-ESP32 library that supports SPI as well as PHY-based Ethernet boards).

All you need to do is establish the connectivity using the Arduino-ESP32 `ETH.begin()` command (with parameters suitable for your board) and HomeSpan will take care of the rest.  The new code is now available in the dev branch if you'd like to test.  Will be released into production as part of HomeSpan 2.1.0 sometime in the next few weeks or so.

I've tested on both the Silicognition WESP32 board (that integrates the PHY interface with the ESP32's built-in MAC interface) as well as a standalone Wiznet W5500 Ethernet board from Adafruit that contains both the PHY and MAC interfaces and communicates with the ESP32 over the standard SPI bus.

Let me know what you think if you have a chance to test.

**Reply by nzim007** [reply]
When will homeSpan 2.1.0 be released?

**Reply by HomeSpan** [reply]
Sometime in next month.  But you should be able to use the dev branch for testing until then.

**Reply by nzim007** [reply]
Can I have the homespan dev library?

**Reply by HomeSpan** [reply]
It's on the HomeSpan GitHub site - you'll need to download it directly, unzip, and overwrite the existing HomeSpan library in your folder.

**Reply by nzim007** [reply]
Ok Thanks
